core/unix/memquery_test.h

```diff
@@ -0,0 +1,344 @@
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/*
+ * memquery_test.h - unit test for unix memquery logic, it is meant to be
+ * included into memquery.c after its includes.
+ *
+ * See also memquery_test_data.h
+ */
+
+#ifdef STANDALONE_UNIT_TEST
+
+typedef struct {
+    memquery_iter_t iter_result;
+    bool is_header;
+    app_pc mod_base, mod_end;
+} fake_memquery_result;
+
+typedef struct {
+    const char *test_name;
+    fake_memquery_result *iters;
+    size_t iters_count;
+
+    const char *in_name;
+    app_pc in_start;
+
+    int want_return;
+    app_pc want_start;
+    app_pc want_end;
+    const char *want_fulldir;
+    const char *want_filename;
+} memquery_library_bounds_test;
+
+#    include "memquery_test_data.h"
+
+/* The fake iterator state during the unit_test_memquery is held in these two
+ * variables; we assume single-threaded usage.
+ */
+static const memquery_library_bounds_test *cur_bounds;
+static int next_iter_position = -1;
+
+static bool
+run_single_memquery_test(const memquery_library_bounds_test *test);
+
+/* unit_test_memquery executes a standalone unit test on
+ * memquery_library_bounds_by_iterator; it is not static because it is needed in
+ * unit_tests.c
+ */
+void
+unit_test_memquery()
+{
+#    ifndef X64
+    // Instead of putting effort into generating test cases which work under both
+    // 32/64 bit, just skip the test when not running 64 bits.
+    return;
+#    endif
+    bool all_passed = true;
+    print_file(STDERR, "START unit_test_memquery\n");
+    for (int i = 0; i < NUM_MEMQUERY_TESTS; i++) {
+        memquery_library_bounds_test *test = &all_memquery_tests[i];
+        print_file(STDERR, "Run memquery unit test %s...\n", test->test_name);
+        const char *status = "OK";
+        if (!run_single_memquery_test(test)) {
+            status = "FAILED";
+            all_passed = false;
+        }
+        print_file(STDERR, "**** memquery unit test %s %s\n", test->test_name, status);
+    }
+    if (!all_passed) {
+        print_file(STDERR, "unit_test_memquery had failing test(s)!\n");
+        os_terminate(NULL, TERMINATE_PROCESS);
```

Comment:
How about just `EXPECT(all_passed, true)` (maybe instead of the whole `if(){}`) so our test failures are all using the same code (for consistency; so we can change from os_terminate or whatnot more easily in the future; etc.)

Reply:
Done.